File: client/src/main/java/Business/Connection/RestConnect.java
package Business.Connection;


import Acquaintence.ConnectionState;
import Business.Interfaces.IParameters;
import Business.Models.Login;
import com.google.gson.Gson;
import org.apache.http.*;
import org.apache.http.client.HttpClient;
import org.apache.http.client.entity.UrlEncodedFormEntity;
import org.apache.http.client.methods.*;
import org.apache.http.client.utils.URIBuilder;
import org.apache.http.entity.ContentType;
import org.apache.http.entity.StringEntity;
import org.apache.http.impl.client.HttpClientBuilder;
import org.apache.http.message.BasicNameValuePair;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.UnsupportedEncodingException;
import java.lang.reflect.Type;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.*;

import static org.apache.http.protocol.HTTP.USER_AGENT;

public class RestConnect {

    private final Gson gson = new Gson();

    private final HttpClient client;
    private final HttpRequestBase request;
    private final PathEnum path;

    RestConnect(PathEnum path, String token, String host, Object content, Object route, IParameters parameters, HttpClient client, HttpRequestBase request) throws URISyntaxException {
        this.client = client;
        this.request = request;
        this.path = path;


        URIBuilder uriBuilder = new URIBuilder(host);
        if (route != null) uriBuilder.setPath(path.getPath() + route.toString());
        else uriBuilder.setPath(path.getPath());

        if (parameters != null) uriBuilder.setParameters(parameters.getParameters());

        this.request.setURI(uriBuilder.build());

        if (content != null) {
            if (!(this.request instanceof HttpEntityEnclosingRequest)) throw new IllegalArgumentException("Request is not of type HttpEntityEnclosingRequest");

            this.request.setHeader(HttpHeaders.CONTENT_TYPE, path.getContentType().getMimeType());
            this.request.setHeader(HttpHeaders.CONTENT_ENCODING, Consts.UTF_8.name());

            if (path.getContentType() == ContentType.APPLICATION_JSON) {
                ((HttpEntityEnclosingRequest) this.request).setEntity(new StringEntity(gson.toJson(content), Consts.UTF_8));
            } else if (path.getContentType() == ContentType.APPLICATION_FORM_URLENCODED) {
                if (content instanceof IParameters) {
                    ((HttpEntityEnclosingRequest) this.request).setEntity(new UrlEncodedFormEntity(((IParameters) content).getParameters(), Consts.UTF_8));
                } else {
                    throw new IllegalArgumentException("Content is not of type IParameters");
                }
            } else {
                throw new UnsupportedOperationException();
            }
        } else if (path.getContentType() != null) {
            request.setHeader(HttpHeaders.CONTENT_TYPE, path.getContentType().getMimeType());
            request.setHeader(HttpHeaders.CONTENT_ENCODING, Consts.UTF_8.name());
        }

        if (token != null) {
            this.request.setHeader(HttpHeaders.AUTHORIZATION, "Bearer " + token);
        }

    }

    public <TContent> RequestResponse<TContent> execute() {
        try {
            HttpResponse response = client.execute(request);
            return getResponse(response);
        } catch (IOException e) {
            return new RequestResponse<>(null, ConnectionState.NO_CONNECTION);
        }
    }

    private <TContent> RequestResponse<TContent> getResponse(HttpResponse response) {
        switch (response.getStatusLine().getStatusCode()) {
            case 200:
                System.out.println("Ok, Object returned");
                return getResult(response);

            case 201:
                System.out.println("Object created and return");
                return getResult(response);

            case 204:
                System.out.println("Ok, nothing returned");
                return new RequestResponse<>(null,ConnectionState.SUCCESS);

            case 400:
                System.out.println("Error, bad request");
                return new RequestResponse<>(null,ConnectionState.BAD_REQUEST);

            case 401:
                System.out.println("Unautherized");
                return new RequestResponse<>(null,ConnectionState.UNAUTHERIZED);

            case 404:
                System.out.println("Not found");
                return new RequestResponse<>(null,ConnectionState.NOT_FOUND);

            case 500:
                System.out.println("Internal Sever Error");
                return new RequestResponse<>(null,ConnectionState.SERVERERROR);
            default:
                return new RequestResponse<>(null, ConnectionState.ERROR);

        }
    }

    private <TResult> RequestResponse<TResult> getResult(HttpResponse response) {
        try {
            BufferedReader rd = new BufferedReader(
                    new InputStreamReader(response.getEntity().getContent()));

            StringBuilder result = new StringBuilder();
            String line;
            while ((line = rd.readLine()) != null) {
                result.append(line);
            }
            Type type = path.getResultType();
            System.out.println(result.toString());
            TResult obj = gson.fromJson(result.toString(), type);
            return new RequestResponse<>(obj, ConnectionState.SUCCESS);
        } catch (IOException e) {
            return new RequestResponse<>(null, ConnectionState.ERROR);
        }

    }


}

File: client/src/main/java/Business/Connection/RestConnect.kt
package Business.Connection


import Acquaintence.IToMap
import com.google.gson.Gson
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import com.ibm.icu.impl.Trie2
import org.apache.http.HttpResponse
import org.apache.http.NameValuePair
import org.apache.http.client.HttpClient
import org.apache.http.client.entity.UrlEncodedFormEntity
import org.apache.http.client.methods.*
import org.apache.http.entity.StringEntity
import org.apache.http.impl.client.HttpClientBuilder
import org.apache.http.message.BasicNameValuePair

import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.io.UnsupportedEncodingException
import java.lang.reflect.Type
import java.util.ArrayList
import java.util.HashMap

import org.apache.http.protocol.HTTP.USER_AGENT

class RestConnect {

    private val url = "https://localhost:5001"
    private val formType = "application/x-www-form-urlencoded"
    //private final String url = "https://ptsv2.com";
    private val gson = Gson()
    // /api/values

    fun login(username: String, password: String): String? {

        try {

            val client = HttpClientBuilder.create().build()

            val request = HttpPost("$url/connect/token/")

            val nvps = ArrayList<NameValuePair>()
            nvps.add(BasicNameValuePair("grant_type", "password"))
            nvps.add(BasicNameValuePair("username", username))
            nvps.add(BasicNameValuePair("password", password))
            request.entity = UrlEncodedFormEntity(nvps)
            val response = client.execute(request)



            println("Response Code Login: " + response.statusLine.statusCode)

            val rd = BufferedReader(
                    InputStreamReader(response.entity.content))

            val result = StringBuffer()
            var line = ""
            while ((line = rd.readLine()) != null) {
                result.append(line)
            }

            println(result.toString())

            val json = JsonParser().parse(result.toString()).asJsonObject

            return json?.get("access_token")?.asString

        } catch (e: IOException) {
            e.printStackTrace()
        }

        return null

    }

    operator fun <T, T1> get(path: PathEnum, route: T1, param: HashMap<String, String>, token: String): T? {
        run {
            try {

                val client = HttpClientBuilder.create().build()

                val request = request<T1, HttpGet>(path, route, param, token)

                var response: HttpResponse? = null



                response = client.execute(request)

                println("Response Code Get : " + response!!.statusLine.statusCode)

                val rd = BufferedReader(
                        InputStreamReader(response!!.entity.content))

                val result = StringBuffer()
                var line = ""
                while ((line = rd.readLine()) != null) {
                    result.append(line)
                }

                println(result.toString())

                val type = path.resultType

                return gson.fromJson<T>(result.toString(), type)

            } catch (e: IOException) {
                e.printStackTrace()
            }


        }
        return null
    }

    fun <T, T1, T2> post(path: PathEnum, param: T, content: T1, token: String): T2? {
        try {

            val client = HttpClientBuilder.create().build()

            val request = request<T, HttpPost>(path, param, null, token)
            val postingString = StringEntity(gson.toJson(content))
            request.entity = postingString

            request.addHeader("Content-type", "application/json")
            var response: HttpResponse? = null
            println(request.entity.toString())

            response = client.execute(request)
            println("Response Code Post: " + response!!.statusLine.statusCode)
        } catch (e: IOException) {
            e.printStackTrace()
        }

        return null
    }


    fun <T, T1, T2> delete(path: PathEnum, route: T, token: String): T2? {
        try {

            val client = HttpClientBuilder.create().build()

            val request = request<T, HttpDelete>(path, route, null, token)


            request.addHeader("Content-type", "application/json")
            var response: HttpResponse? = null

            response = client.execute(request)
            println("Response Code Post: " + response!!.statusLine.statusCode)
        } catch (e: IOException) {
            e.printStackTrace()
        }

        return null
    }

    //TODO implement
    fun <T, T1, T2> put(path: PathEnum, route: T, content: T1, token: String): T2? {
        try {
            val client = HttpClientBuilder.create().build()

            val request = request<T, HttpPut>(path, route, null, token)
            val postingString = StringEntity(gson.toJson(content))


            request.entity = postingString

            request.addHeader("Content-type", "application/json")
            var response: HttpResponse? = null
            println(request.entity.toString())

            response = client.execute(request)
            println("Response Code Post: " + response!!.statusLine.statusCode)
        } catch (e: IOException) {
            e.printStackTrace()
        }

        return null

    }

    /***
     *
     * @param path Which call from client
     * @param route The route variable
     * @param param Params for get method (After the ?)
     * @param token The login token
     * @param <T> Generic type of the route parameter
     * @param <T1> The respective request type
     * @return The respective request
    </T1></T> */
    fun <T, T1> request(path: PathEnum, route: T?, param: HashMap<String, String>?, token: String): T1 {
        var request: HttpRequestBase? = null
        var url = this.url + path.path
        if (route != null) {
            url = url + route.toString()
        }
        if (param != null) {
            url = "$url?"
            for ((key, value) in param) {
                url = url + (key + "=" + value)
            }
        }
        //TODO delete this line
        //url = url.concat("?departmenId=1");


        println(url)
        when (path.type) {
            ConnectionType.POST -> {

                request = HttpPost(url)
            }
            ConnectionType.GET -> {
                request = HttpGet(url)
            }

            ConnectionType.DELETE -> {
                request = HttpDelete(url)
            }

            ConnectionType.PUT -> {
                request = HttpPut(url)
            }
        }
        request.addHeader("User-Agent", USER_AGENT)
        request.addHeader("Authorization", "Bearer $token")


        return request as T1?
    }


}
